### PR TITLE
8206187: javax/management/remote/mandatory/connection/DefaultAgentFilterTest.java fails with Port already in use

### PR DESCRIPTION
**Pull request by kevinjwalls**
Test fails occasionally due to a port clash.
Presumably the port that was returned by Utils.getFreePort(), is no longer free.
The test creates a ProcessBuilder with the parameters for JMX, including port number, and uses that to create a new Process.
It should retry with a new port if we fail due to a port in use, for some limited number of attempts.

main already has some retry logic, but not working:
it checks for an InvocationTargetException to contain a BindException, but it simply gets a BindException, thrown by TestAppRun.start().
TestAppRun.start() runs the new process and scans for errors, but on failure its predicate has only seen the first line of a failure, so a BindException is never recognised and thrown.
Also main does not limit the retries, and handling the port retry in main() is duplicated, for each run of the test method.

So...

Make the error-scanning predicate in TestAppRun recognise a "port in use" message and throw a BindExeption.  This is a notification to the caller that it failed, it's not the actual BindException as that was thrown in a different process.
Make the testDefaultAgent method (the main part of the test) handle retrying with a new port, a limited number of times.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8206187](https://bugs.openjdk.java.net/browse/JDK-8206187): javax/management/remote/mandatory/connection/DefaultAgentFilterTest.java fails with Port already in use

### Reviewers
 * [Mark Sheppard](https://openjdk.java.net/census#msheppar) (@msheppar - **Reviewer**)
 * [Alex Menkov](https://openjdk.java.net/census#amenkov) (@alexmenkov - **Reviewer**) ⚠️ Review applies to 6cdcde8aac6a3f2087d34e2287276a964be60279

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7589/head:pull/7589` \
`$ git checkout pull/7589`

Update a local copy of the PR: \
`$ git checkout pull/7589` \
`$ git pull https://git.openjdk.java.net/jdk pull/7589/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7589`

View PR using the GUI difftool: \
`$ git pr show -t 7589`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7589.diff">https://git.openjdk.java.net/jdk/pull/7589.diff</a>

</details>

### CODE REVIEW

**Comment by msheppar** [comment]
This is reasonable approach to take and should ameliorate the BindException failure condition.
But, as it is using the same port allocation strategy which is in current use, then it will be still be
suspect to intermittent BindException failures. This will be especially true, as with the current failures,
there is a lot of network testing activity with concurrent network test execution and the MACH5 test framework itself e.g. logstash etc.

An alternative is to use a fixed port 1098, which previously was the default for RMI Activation daemon.
This is now available as activation has been removed from the java platform in JDK17.

**Comment by alexmenkov** [comment]
The test uses warm-up predicate in a strange way - it returns true for any output from child process.
The main purpose of the predicate is to wait until some expected output is produced, and return false to skip other lines (for example in some environments VM may print some warnings).
For this test it may be "main enter" (printed from TestApp.main()).
To handle "port in use" error the predicate can search for "bindexception" or "port already in use" line.
I don't think check for "exception" and "error" makes much sense here.

**Comment by kevinjwalls** [comment]
> An alternative is to use a fixed port 1098

Thanks Mark - I will avoid that fixed slot as we no doubt run tests concurrently, and also in case these get backported far enough that it's not free. 8-)

Utils.getFreePort() lets new ServerSocket choose a port,  but there's clearly a race to use it.  We could make it simply random, but I think we need still need to retry like this to avoid failures --  will monitor and see how the race goes with 10 attempts.

**Comment by kevinjwalls** [comment]
> The test uses warm-up predicate in a strange way

Thanks Alex - yes so that's why we only see one line in the predicate, it contains the word "exception" and the predicate returns true, signalling that the process is done starting up. 8-)
But if you do see an exception, you're never going to see the text from the app that it has started OK, so I see why it was returning true always - we were either seeing an error, or otherwise we presume TestApp has started.
But it returned true after seeing the first line, which does not contain "bindexception", so it never saw the BindException later to recognise the port being in use.

Anyway...
I will update shortly to hopefully clarify the predicate part.  Still good to keep main() simpler and not have duplicated retry logic there.

File: test/jdk/javax/management/remote/mandatory/connection/DefaultAgentFilterTest.java [hunk 1]
```diff
@@ -164,9 +169,24 @@ public void close() throws Exception {
     }
 
     private static final String TEST_APP_NAME = "TestApp";
+    private static final int FREE_PORT_ATTEMPTS = 10;
 
     private static void testDefaultAgent(String propertyFile) throws Exception {
```
Comment (msheppar) [inline test/jdk/javax/management/remote/mandatory/connection/DefaultAgentFilterTest.java:174]: This is a good refactoring change as it also removes some broken logic in the current retry strategy

**Review by msheppar — APPROVED** [review]

**Review by alexmenkov — APPROVED** [review]

**Comment by kevinjwalls** [comment]
Thanks for the reviews!

Apologies, one more update - I had a break in there earlier to terminate the retry loop, and I removed it accidentally when changing things to make it fail.
The retry loop has to let the testDefaultAgent method complete normally (which needs the break), or by throwing an Exception.
Both looking good now.

**Comment by msheppar** [comment]
A good spot by yourself ... as the focus was on the BindException correction and the pass condition is an Exception being thrown, it was easy to miss the break for a failure condition ... a similar issue existed in the current test albeit only if an InvocationTargetException with a nested BindException was thrown and the retry resulted in the 
InvalidClassException or UnmarcshalledException !!

**Review by msheppar — APPROVED** [review]

**Pull request closed without merging**